internal/expr/prepare.go

```diff
@@ -0,0 +1,183 @@
+	s := 0
+	for _, fn := range fns {
+		if fn.name == "*" {
+			s++
+		}
+	}
+	return s
+}
+
+func prepareOutput(ti typeNameToInfo, p *outputPart) ([]string, error) {
+
+	var outCols []string = make([]string, 0)
+
+	// Check target struct type and its tags are valid.
+	for _, t := range p.target {
+		inf, ok := ti[t.prefix]
+		if !ok {
+			return nil, fmt.Errorf("unknown type: %s", t.prefix)
+		}
+
+		_, ok = inf.tagToField[t.name]
+		if !ok && t.name != "*" {
+			return nil, fmt.Errorf(`there is no tag with name "%s" in "%s"`, t.name, inf.structType.Name())
+		}
+	}
+
+	// Check asterisk are in correct places.
+
+	sct := starCount(p.target)
+	scc := starCount(p.source)
+
+	if sct > 1 || scc > 1 || (scc == 1 && sct == 0) {
+		return nil, fmt.Errorf("invalid asterisk in output expression")
+	}
+
+	starTarget := sct == 1
+	starColumn := scc == 1
+
+	lenS := len(p.source)
+	lenT := len(p.target)
+
+	if (starTarget && lenT > 1) || (starColumn && lenS > 1) {
+		return nil, fmt.Errorf("invalid mix of asterisk and none asterisk columns in output expression")
+	}
+
+	if !starTarget && (lenS > 0 && (lenT != lenS)) {
+		return nil, fmt.Errorf("mismatched number of cols and targets in output expression")
+	}
+
+	// Case 1: Star target cases e.g. "...&P.*".
+	// In parse we ensure that if p.target[0] is a * then len(p.target) == 1
+	if starTarget {
+
+		inf, _ := ti[p.target[0].prefix]
+
+		// Case 1.1: Single star e.g. "t.* AS &P.*" or "&P.*"
+		if starColumn || lenS == 0 {
+			pref := ""
+
+			// Prepend table name. E.g. "t" in "t.* AS &P.*".
+			if lenS > 0 && p.source[0].prefix != "" {
+				pref = p.source[0].prefix + "."
+			}
+
+			for tag := range inf.tagToField {
+				outCols = append(outCols, pref+tag)
+			}
+
+			// The strings are sorted to give a deterministic order for
+			// testing.
+			sort.Strings(outCols)
+			return outCols, nil
+		}
+
+		// Case 1.2: Explicit columns e.g. "(col1, t.col2) AS &P.*".
+		if lenS > 0 {
+			for _, c := range p.source {
+				if _, ok := inf.tagToField[c.name]; !ok {
+					return nil, fmt.Errorf(`there is no tag with name "%s" in "%s"`,
+						c.name, inf.structType.Name())
+				}
+				outCols = append(outCols, c.String())
+			}
+			return outCols, nil
+		}
+	}
+
+	// Case 2: None star target cases e.g. "...&(P.name, P.id)".
+
+	// Case 2.1: Explicit columns e.g. "name_1 AS P.name".
+	if lenS > 0 {
+		for _, c := range p.source {
+			outCols = append(outCols, c.String())
+		}
+		return outCols, nil
+	}
+
+	// Case 2.2: No columns e.g. "&(P.name, P.id)".
+	for _, t := range p.target {
+		outCols = append(outCols, t.name)
+	}
+	return outCols, nil
+}
+
+// Prepare takes a parsed expression and all the Go objects mentioned in it.
+// The IO parts of the statement are checked for validity against the Go objects
+// and expanded if necessary.
+func (pe *ParsedExpr) Prepare(args ...any) (expr *PreparedExpr, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot prepare expression: %s", err)
```

Comment:
I think I recall seeing this in an older PR, it still seems a little magical to me (and hence un-Go-like?)

Would it be cleaner to have a function to do this (name tbd) and to call it explicitly each time?
```go
func (pe *ParsedExpr) Prepare(...) (...) {
    errFmt := func(e error) error {
        return fmt.Errorf("cannot prepare expression: %s", e)
    }
    ...
        return nil, errFmt(err)
}
```

Reply:
It will, since both parse and prepare are actually in the internal subdirectory we will have a nice user friendly wrapper round the outside that brushes over confusing names like parse prepare and complete. But thats all for the future.